### ISSUE TITLE
Unpack RAR files

### ISSUE THREAD

**Issue by markus101** [Sonarr/Sonarr#784]
This would help out with RARed torrents and make the fact that torrent clients don't have a post processing state less problematic (Sonarr won't try to import a partially unpacked file).

**Comment by subzero79** [comment]
I think this would be a problem being handled by sonarr in case of remote downloaders. IMO the client should always handle the unpacking.
Would it be possible to use the sonarr api in the unpacking script after unraring, to do a manual import with delete (copy and delete source) from the path? If we know the output file name send it to sonarr to look for it.

**Comment by subzero79** [comment]
Would it be possible to add a env var to custom scripts with the import path (local or remote) of the imported file, this way we can add a custom script to remove the file  (mkv | mp4 | avi) if rar files are present or checking the files against the client with rpc. I am thinking in bash here.

I've just moved from deluge to transmission because of my low spec vps, just realised the sync folder to drone factory doesn't work anymore, the queue at sonarr get's stuck, unless i move the path (remove the label). It used to work in deluge

**Comment by qupncy** [comment]
couchpotato handles this instead of being done at the downloaders end.. Not familiar with the ins and outs of how it is done there but maybe it is something you guys can implement into sonarr?

**Comment by subzero79** [comment]
Some corrections were introduced to the env vars of sonarr for custom post processing. Also it was added source path and source file name

With this vars we will be able to execute scripts after import that can clean if the source import contains rar's, deleting the source file.

This is with the latest commits, but a built hasn't been pushed yet to try how it works in real.

**Comment by qupncy** [comment]
Ah sounds good.. if possible, it might be a good idea to give users options to keep the source file (for seeding) purposes if it was downloaded via torrent..

**Comment by subzero79** [comment]
That's the idea, your removing only the video file, not the rar's that are seeding.

And it will only perform video file deletion in case it contains rar's.

**Comment by subzero79** [comment]
Just an update on this, the latest develop version updated the vars, i tested the script to remove the source file in case rar's exists. Everything good so far. 

This was done in bash, the scripts are passed in lowercase (i got confused by the code and the wiki) in case someone is interested. 

example of the important vars for this case

$sonarr_episodefile_sourcepath=/mnt/remote/torrent/finish/tv-sonarr/Penny.Dreadful.S01E01.720p.HDTV.x264-2HD/penny.dreadful.s01e01.720p.hdtv.x264-2hd.mkv
$sonarr_episodefile_sourcefolder=/mnt/remote/torrent/finish/tv-sonarr/Penny.Dreadful.S01E01.720p.HDTV.x264-2HD

**Comment by qupncy** [comment]
I'm pretty confused myself. I'm looking for sonar to extract the torrent (when finished) from D:\Programs and save the extracted file to E:\Completed\Dronefactory

Any tips on what i need to put in my script so i can test for you guys?

**Comment by subzero79** [comment]
Sonarr won't extract. Any torrent client can extract. 

The idea is that the download client extracts to the same folder, the sonarr would find the extracted video file, CDH will copy, and once the file has been imported the script will look at the original source folder for rar files, if they are present it will delete the source file used for import, leaving the rar's.

**Comment by Taloth** [comment]
Also listed in the torrent part 2 checklist https://github.com/Sonarr/Sonarr/issues/468.

**Comment by markus101** [comment]
Yeah, I was trying to get away from the meta issues, but I should link it back here.

**Comment by Taloth** [comment]
I just wanted to mention it, so if someone picks up this issue he'll know to check off the other one too.

**Comment by duelistjp** [comment]
would love to see this as well or at least a message from sonarr that there is a rar file that matches a downloaded episode so i know without having to go looking through everything

**Comment by iheijoushin** [comment]
I also would love to see Sonarr have the ability to UnRAR, as this is the only thing I miss about SickRage. 

I use a remote Seedbox running rTorrent, which hard-links completed files to folder that is synchronized back to my NAS using BTSync, which then drops the files into the watch folder that Sonarr uses.

While I *could* have rTorrent unRAR on the Seedbox (a process that is pretty unreliable unto itself), I then have to move a much larger file which takes longer/eats up bandwidth, and brings me closer to my ISPs download limit.

Please consider it! :)

**Comment by Taloth** [comment]
@iheijoushin Most (if not all) rars aren't actually compressed, only packed.
Because rar compression of a videofile won't yield any additional compression, if it did, you wouldn't need a special video codec.

**Comment by iheijoushin** [comment]
Interesting! So my argument for bandwidth may not be entirely accurate. However, my experience with rTorrent's post-processing has not been great...

Sonarr works spectacularly well with the features it offers, and I hope UnRARing is added in the future so we can be spared lesser software. :)

**Comment by Fidelix** [comment]
@subzero79 can you show me this bash script you're talking about, and how you're using it?

Thanks.

**Comment by subzero79** [comment]
@Fidelix 
Yes sure, i don't use this setup anymore, which was cleaning the source in a remote path over NFS.
I use syncthing now.

https://gist.githubusercontent.com/subzero79/87a347a07964390884c9/raw/6458263ac9b2be8caf6cf7218c24dcd505be67a9/Clean%2520source

This script assumes rar types of r01/r02 etc. I've seen other types like part01.rar part02.rar which for this script would have to be redone or use one rar file check

**Comment by Fidelix** [comment]
@subzero79 thank you for sharing.
Can you describe how you use syncthing in a setup for TV shows? So you just duplicate all the data?

**Comment by subzero79** [comment]
The downloader is on a remote site (transmission,vps).
When files are completed all of them are copied to an export folder (syncthing), let's say `/storage/export/series`
The copying/moving process varies according to the type of torrent:, but they are all processed through transmission torrent-done-script
Torrent with folder and video file get's hard linked to a folder with the same torrent folder name to export
Torrent without folder and single video file, get's hard linked to export
Torrent with folder and rars: video file gets extracted to the same torrent folder and then moved to export inside a folder with the same name as the torrent.

At the sonarr side, the system will be receiving file from syncthing, in sonarr you have to use remote path to translate the path sonarr is expecting. When a syncthing process happens sonarr will scan the folder several times and you can see it in the logs, no files found at path, this is because files have a tmp extension plus a .syncthing prefix. When the sync is finished and files have no longer the temp extension sonarr will import. From there you need a post processing script to remove the folder, which will also happen to remove the remote folder at export folder. Files continue to be seeding

**Comment by hmemcpy** [comment]
Hey, I know this is an old issue, but having discovered Sonarr recently (and already having contributed [Vuze support](#1277), I was thinking to add this missing piece of the puzzle.
Some trackers I use insisting on using .rar files still, and in the past I had some automation in Vuze to unpack them, but it still left some maintenance issues to deal with.

I was thinking about approaching handling archives in Sonarr naively by abstracting the archive away, possibly using `IDiskScanService` that could retrieve video files directly from the archive. Using a library such as [SharpCompress](https://github.com/adamhathcock/sharpcompress) to read the contents of the archive, and doing the extracting/cleanup. Then the various import services would operate on those files directly.
It would still need to extract the files to a temp location, but that would be the compression library's business.

How's that sound?

**Comment by kayone** [comment]
Some questions so we can clarify somethings before starting to write da codez:

- is it safe to assume it would only be used with torrents? any usenet client that don't do this?
- do we want to be smart about when this kicks in? eg. only for download clients that we know don't support unziping? or do we assume a zip file is always possible?
- as for the temp folder, do we need one? can't we decompress directly to target folder?

**Comment by Taloth** [comment]
`IDiskScanService`, it's not that easy.

Some issues:
- Import decisions using the content (sample check uses mediainfo)
- Remote systems/seedboxes
- Files not being imported end up being decompressed uselessly.
- rar sets containing season packs, import decisions run per video.

Most of these issues are solvable, but there are quite a few edge cases to consider.

For transfer providers (https://github.com/Sonarr/Sonarr/issues/282) I wanted to create some kind of virtual abstraction of the file system with specific features that would allow indexing and import/copy/move to be performed separately.

**Comment by hmemcpy** [comment]
> For transfer providers (#282) I wanted to create some kind of virtual abstraction of the file system with specific features that would allow indexing and import/copy/move to be performed separately.

I was thinking the same: instead of returning string paths, provide abstractions over the file entries, so they can accept (a-la visitor pattern) the import decision making implementations. These abstractions can represent a media file on disk or a media file inside an archive...

I'll try spiking something, will send a PR to discuss...

**Comment by RobertDeRose** [comment]
I didn't read the entire thread here, but in case anyone else is using Transmission with Sonarr, this is my very basic script to handle unraring.  

https://gist.github.com/RobertDeRose/9391f5da6273eab26f00d2e7c3c945d3

Just modify the `unrar_staging` directory and tell Transmission to run this script when downloads are complete.

**Comment by painejake** [comment]
Must admit I just have a cron job setup for every minute:

`find /home/transmission/complete/ -name '*.rar' -execdir unrar e -o- {} \;`

Seems to handle everything I've encountered and its nice and simple

**Comment by RobertDeRose** [comment]
@painejake does that handle not unrar`ing the same file more than once? I can't find a reference for the `-o-` on the googles. :(

**Comment by n-shay** [comment]
Not sure if this is the right place to publish it but for now it will do.
If you are using Sonarr and Transmission like me I found a solid workaround solution for this pending feature.

Here are the a few simple steps:

1. Configure Transmission to run a script at `script-torrent-done-filename`: [transmissionUnRAR.sh](https://gist.github.com/n-shay/ddf81b12afbe295190cca5f7f929dca1) (forked from @marcocot, thanks!)
2. If needed, modify the `DIR_CONTAINS` variable to the name of the folder Sonarr is downloading files to.
3. On Sonarr, make sure you are using hardlinks: Settings -> Media Management -> Importing -> "Use Hardlinks instead of Copy" is set to Yes.
4. Add a cleanup script to Sonarr: Settings -> Connect -> Add . Check "On Download" and "On Upgrade", and specify the path to the local script path: [sonarr_cleanup.php](https://gist.github.com/n-shay/ff8b1c0fc009c947b1bf7772e500e763).

** Don't know if it's absolutely necessary, but you might want to give execute permissions to both scripts.

Both scripts will create log files at `/var/log/` to let you know if they're working and what they're doing.

**Comment by seanvree** [comment]
does anyone have a solution for deluge?

**Comment by onigoetz** [comment]
Hi @seanvree, for Deluge you can use the extractor plugin, but it doesn't work in conjunction with "move on completed" (it unrars where the file was before it was moved ... not ideal.)

For my usage, I made a small cron that will scan a folder for all .rar files, list their content and if the files aren't extracted yet, extract them.

It seems to work perfectly fine.
If you want I can post it in a gist ?

**Comment by seanvree** [comment]
@onigoetz , I"m aware of the plugin, but like you said, it doesn't work if you have labels that sort the files into other folders, which I do.

Will the CRON job run on Windows?

**Comment by onigoetz** [comment]
it might with bash on windows, but not natively.

**Comment by michaelarnauts** [comment]
You light be able to use the task scheduler on Windows for this.

**Comment by seanvree** [comment]
@onigoetz I'll try it! Hook me up!

**Comment by onigoetz** [comment]
I'm sorry it took so long, here's the gist : https://gist.github.com/onigoetz/277df51931d2e6acdcee9dedd42fdf73

**Comment by seanvree** [comment]
@onigoetz THANKS!

Noob question tho since I have BASIC knowledge of linux code. 

Where do I set the path of the directory to look for .rar files??

I was able to get the script running in windows, but I get this error:


C:\SonarrUnrar>sh unrar.sh
Starting UNRAR
unrar.sh: line 27: syntax error near unexpected token `<'
unrar.sh: line 27: `done < <(find $DIR -name '*.rar' -print0)'

Thanks

**Comment by onigoetz** [comment]
You're welcome

I see here, you run this scripts with `sh unrar.sh` you have to use `bash` not `sh`.

Then, to define the path, it's the first parameter to the script, for you that would be something like : 
`bash unrar.sh "C:\Data\"`

**Comment by seanvree** [comment]
@onigoetz , OKAY! Making progress...

So i'm testing with a directory that has a .rar file in it...so this is what my commands are:


C:\WINDOWS\system32>cd C:\SonarrUnrar

C:\SonarrUnrar>bash unrar.sh
Starting UNRAR
File not found - *.rar
UNRAR Done

C:\SonarrUnrar>bash unrar.sh "C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS"
Starting UNRAR
File not found - *.rar
UNRAR Done

C:\SonarrUnrar>bash unrar.sh "C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.rar"
Starting UNRAR
File not found - *.rar
UNRAR Done

C:\SonarrUnrar>


However, there is a .RAR file in that directory....

what am I doing wrong?

**Comment by onigoetz** [comment]
I have no idea, you're doing it correctly from what I see.
I'm not very familiar with bash on windows.

Maybe the paths are different ? I see on your screenshot that bash sees the path as `/cygdrive/c/...`

Try to run the command like this : `bash unrar.sh "/cygdrive/c/plexdownloads/TV"`

**Comment by seanvree** [comment]
@onigoetz , okay, I'm making huge progress here! Really appreciate you helping I"m assuming this work will be able to thelp out @Taloth  when the implment this in Sonarr.

So, first the problem with it not finding the directly, I added this line to top the top of the .sh file:

"export PATH=/bin/:/usr/bin:$PATH"

That line is required for windows. 

then, ran this command:

bash unrar.sh "/cygdrive/c/media/plexdownloads/TV/Shark.Tank.S08E07.720p.HDTV.x264-CROOKS"

And it finally found the file and tried to extract it, but it's having a hard time with the UNRAR command in the .SH file. 

so the script is passing the UNRAR command, but it's not passing the right switch. You have X which it should work.

So, I  manually ran the UNRAR command on the same file, and it worked:

C:\SonarrUnrar>unrar x C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.rar

UNRAR 5.40 x64 freeware      Copyright (c) 1993-2016 Alexander Roshal


Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.rar

Extracting  shark.tank.s08e07.720p.hdtv.x264-crooks.mkv                5%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r00

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               11%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r01

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               16%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r02

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               22%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r03

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               27%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r04

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               33%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r05

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               38%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r06

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               44%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r07

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               49%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r08

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               55%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r09

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               61%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r10

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               66%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r11

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               72%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r12

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               77%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r13

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               83%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r14

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               88%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r15

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               94%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r16

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               99%

Extracting from C:\Media\plexdownloads\TV\Shark.Tank.S08E07.720p.HDTV.x264-CROOKS\shark.tank.s08e07.720p.hdtv.x264-crooks.r17

...         shark.tank.s08e07.720p.hdtv.x264-crooks.mkv               OK
All OK


SO, windows is having a hard time passing this line:

"  unrar x $_RAR $_DIRNAME "

Any ideas?

**Comment by onigoetz** [comment]
I'm not an expert in bash scripting either, I mostly use stackoverflow to create my bash scripts :p 

What's weird is that  unrar seems to like the path with `C:\...` but bash passes the path as `/cygdrive/c/...` Try to run `bash` and inside the prompt, run `unrar x ` with the path in unix style

**Comment by seanvree** [comment]
so I ran  "unrar x /cygdrive/c/media/plexdownloads/TV/Shark.Tank.S08E07.720p.HDTV.x264-CROOKS"

and it didn't extract it.  Soooo.....unrar has to use the C:\ method.

**Comment by seanvree** [comment]
any ideas @onigoetz ?

**Comment by onigoetz** [comment]
I'm sorry, I have no idea how to fix this

**Comment by seanvree** [comment]
@onigoetz , NO problem mate, looks to be too much for Windows! Thanks brother!

**Comment by seanvree** [comment]
FYI!!! Found an easy and fully automated solution:

https://forums.sonarr.tv/t/short-easy-tutorial-on-how-to-auto-extract-tv-shows-when-using-windows-with-no-scripts-service/12900

**Comment by meffect** [comment]
need some flock action going on here

**Comment by lfmnovaes** [comment]
@seanvree Hey, can u provide the full script u made a while ago? I really want a nice way to do it, and maybe i can help.

**Comment by seanvree** [comment]
I never wrote a script. I wish.  I never found a good script.  If someone can make a script out of this, that would be coo, I'm a bomb windows coder, but not a good scripter:

https://forums.sonarr.tv/t/guide-automatically-process-rar-releases-in-sonarr-radarr-while-using-usenet-and-torrents-on-a-windows-environment/13743?source_topic_id=12900

**Comment by ElevatorCh** [comment]
This is a necessary feature for a lot of us. Would love to see it implemented!

**Comment by Bokaii** [comment]
+1 vote for this feature. It is the only thing that keeps me from switching to Sonarr(and Radarr).

**Comment by seanvree** [comment]
All, I worked with @Taloth to put together a VERY effective and simple solution to this that DOES NOT use the drone factory and only requires one free small OSS package. 

Please see the post below:

https://forums.sonarr.tv/t/unrar-win-new-guide-to-auto-process-rar-torrents-via-sonarr-w-o-drone-factory/15381

**Comment by Bokaii** [comment]
Great work, @seanvree, and beautiful guide. However this line makes me sad "And if you’re using Linux, please close this page now."  :(

I am using a Synology NAS and have been trying for two days to get a good unrar-script to work well. This process would all be so much easier if Sonarr could do the unpacking.

**Comment by Anthirian** [comment]
I agree. Although the tutorial posted by @seanvree is nice, it still remains a workaround.

**Comment by seanvree** [comment]
@Bokaii  - ha, as I understand it, there are many solutions to unrar torrents using linux, however, I cannot confirm as I don't use linux...obv. I wrote that for Windows because until that was authored there was not a single solution.

@Anthirian  - Agreed.  I DO KNOW that there is work to bake UNRAR into sonarr....so until that happens, that's all we've got.

**Comment by ElevatorCh** [comment]
Still looking into this for Linux. If anyone have a solution, please let us know!

**Comment by jyggen** [comment]
@ElevatorCh: I use [this script](https://techblog.jeppson.org/2016/11/automatically-extract-rar-files-downloaded-transmission/) with some minor changes.

**Comment by ElevatorCh** [comment]
@jyggen Thanks! I'll look into it 👍

**Comment by n-shay** [comment]
@ElevatorCh which DL client do you use? 

I got a pretty solid solution for [transmission ](https://gist.github.com/n-shay/ddf81b12afbe295190cca5f7f929dca1)with [sonarr](https://gist.github.com/n-shay/ff8b1c0fc009c947b1bf7772e500e763)/[radarr](https://gist.github.com/n-shay/b03bc712e9c58eb5583ba6bb69ad3ec7).

I wrote a few steps on this [thread before](https://github.com/Sonarr/Sonarr/issues/784#issuecomment-256170125).

**Comment by ElevatorCh** [comment]
@n-shay I use Deluge, but I think I'll adjust the scripts you guys sent me and do an API call to sonarr when extraction is complete. That seems to be the best way to do it according to everything I've read so far, and a good way to avoid partially finished extractions to be grabbed.

**Comment by Chris-V** [comment]
hey guys, I recently got pretty tired of manually ssh'ing into my server to unrar files, so I wrote this small python script to do it automatically.

So for anyone on Linux and using Deluge, you could get along with this: https://gist.github.com/Chris-V/cfe83564d62da0ff2bc0f3f82e5ebb49

It's not perfect but at least extractions are automated now :)

**Comment by levic92** [comment]
@seanvree @onigoetz @ElevatorCh @Chris-V 

I have made my own plugin for Deluge to better support Sonarr and Radarr, feel free to try it out. It has been working great for me. I am toggling torrent properties so that it doesn't get imported while the extraction is happening.

https://github.com/levic92/LCExtractor

**Comment by seanvree** [comment]
@levic92 ! GAME CHANGER! Checking it out now!

**Comment by bardisty** [comment]
@levic92 works great! Thank you!

**Comment by ElevatorCh** [comment]
@levic92 Perfect! Works exactly how I want it too!

Now to get sonarr to delete the extracted file after import... hmmmmmmmmmm.....

**Comment by seanvree** [comment]
@ElevatorCh  sonarr and radarr both work with the plugin.

**Comment by levic92** [comment]
@ElevatorCh give this a try to delete extracted files [https://github.com/levic92/LCExtractor/tree/master/extras](url)

**Comment by DemonCow5761** [comment]
Years later and I'm still dealing with this as a torrent user :-(  SimpleExtractor for deluge doesn't work all the time, LCExtractor for deluge doesn't work all the time, the default extractor that comes with deluge doesn't work hardly ever.  Extract now isn't great either n leaves lots of leftover stuff and seems like a resource hog.

I just want full automation after so long of trying different methods I just been using LCExtractor since that came out but I still have to manually extract several times a week usually.

TLDR
Please add unrar, still very much needed for many of us.

**Comment by 0xdevalias** [comment]
Even if this wasn't added as an 'automagic' feature, it would be pretty cool to have on the Activty -> Manual Import screen.

Where it says the following:
> No video files were found in the selected folder.

Perhaps it could detect the presence of basic (or even better, configurable) archive types: zip, rar, etc; and then provide an extract button.

The UI parts seem to be here, but don't know enough about the codebase to locate the backend bits off the top of my head:
* https://github.com/Sonarr/Sonarr/tree/master/src/UI/ManualImport
* https://github.com/Sonarr/Sonarr/blob/master/src/UI/ManualImport/EmptyViewTemplate.hbs

**Comment by jkaberg** [comment]
For anyone looking for a solution there is this; https://github.com/jkaberg/tExtractor

It's been rock solid for me for well over a year now. This solution solves the nested archive issues wich some torrents have.

**Comment by ThePieMonster** [comment]
Currently when downloading a torrent that contains .rar files the download will show that it's almost complete but not 100% because it is not able to extract the .rar files. The only way to progress here is to navigate to the torrent location and extra the files manually. Is there not a check that can be put in place to auto-extract the files on torrent download completion? 

![image](https://i.postimg.cc/fRJC58dB/Untitled.png)

**Comment by bakerboy448** [comment]
https://github.com/davidnewhall/unpackerr  exists

this likely won't occur anytime soon, so please do not ask when it will be done.

**Comment by daburger** [comment]
> https://github.com/davidnewhall/unpackerr exists
> 
> this likely won't occur anytime soon, so please do not ask when it will be done.

Yeah well, it still needs some changes I think.

I have been using unpacker for a while and it works...kinda. My torrents that are incomplete are kept on my SSD until they finish. Once they are finished downloading, they move to a folder located on a different share (SSD not part of this share) where they sit and seed. These downloads are seeding indefinitely (because I am a bloody champion) and so I configured all of this to work with hardlinks. For those who don't know what hardlinks are, they basically allow me to make a shortcut-like-file of the downloaded torrent to my media folder without having a copy (so no need to have a copy which takes up double the space for no reason). This allows me to continue seeding.

The problem I have is that Unpacker will not unpack any rar files because those that are seeding are not considered as "complete" by sonarr. So I have to manually unpack them anyway.

It would be nice to have an option to unpack regardless if the download is not complete. This would solve my problem. I tried using the folder option as oppose to linking Sonarr and Radarr but that didn't work as it doesn't seem to monitor subfolders.

back to square 1

**Comment by jeffbyrnes** [comment]
@daburger so I have a setup similar to yours:

* Find things via the -arr apps
* Use qbittorrent to do downloads to `/volume/data/downloads`
* Use Unpackerr to handle RAR files once download is finished
* Hard linking extracted files to my media library location on same volume

It works well, and a torrent being “downloaded” but still seeding doesn’t prevent Unpackerr from operating.

Do you have the various config options set up so Unpackerr can read the various -arr apps’ APIs and see when a download is ready to be unarchived?

**Comment by Nicxx2** [comment]
@daburger 

You can also use this:
https://github.com/Nicxx2/auto-unrar

It is an easy setup if you are using docker, or you can just use the script (just edit the source directory path, which is the starting point for the search process to extract RAR files).

Basically it will automatically extract RAR files, it won't delete them or overwrite (in case files were extracted before). Also creates a mark file within the folders that are already extracted, to skip quickly and go to the next ones.

**Comment by edrock200** [comment]
check out rar2fs as well. Mounts your torrent dir, and presents rars/compressed files as the actual. Unpacks on the fly when copied out. No need to unrar.
